src/main/org/deidentifier/arx/dp/ExponentialMechanism.java

```diff
@@ -14,111 +14,147 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-
 package org.deidentifier.arx.dp;
 
-import java.math.BigInteger;
+import java.math.BigDecimal;
+import java.math.MathContext;
+import java.math.RoundingMode;
 import java.security.SecureRandom;
-import java.util.HashMap;
-import java.util.Map;
+import java.util.ArrayList;
+import java.util.List;
 import java.util.Random;
-
-import org.apache.commons.math3.fraction.BigFraction;
+import org.apache.commons.math3.analysis.function.Exp;
+import org.apache.commons.math3.distribution.EnumeratedDistribution;
+import org.apache.commons.math3.random.AbstractRandomGenerator;
 import org.apache.commons.math3.util.Pair;
-import org.deidentifier.arx.reliability.IntervalArithmeticDouble;
-import org.deidentifier.arx.reliability.IntervalArithmeticException;
 
 /**
- * An implementation of the reliable variant of the exponential mechanism.
+ * An implementation of the exponential mechanism for discrete domains as proposed in:
+ * 
+ * McSherry, Frank, and Kunal Talwar:
+ * Mechanism design via differential privacy.
+ * Foundations of Computer Science 2007. pp. 94-103
+ * 
  * This implementation assumes that all score values have been divided by the sensitivity of the respective score function.
  * 
- * Note: This implementations uses internal caches which may grow up to the size of distinct values
- * provided as scores to the method setDistribution during the lifetime of an instance.
+ * We point out that this implementation draws from a probability distribution which approximates the mathematically precise
+ * distribution of the exponential mechanism as a consequence of floating-point arithmetic, which could potentially affect
+ * the privacy guarantees provided. However, it can be shown that the resulting exceedance of the privacy parameter epsilon
+ * is at most in the order of log((1 + n * 10^{1-p}) / (1 - n * 10^{1-p})) where n is the size of the domain and p is
+ * the floating-point precision which can be specified by the user.

```

Comment:
I see doubles being used below. Are you sure that we are not restricted by the precision of doubles here? What would that mean for the error described in the formula above?